rp2040-hal/examples/gpio_dyn_pin_array.rs

```diff
@@ -0,0 +1,120 @@
+//! resistor on each of GPIO 2, 3, 4, 5.  
+//! The other side of the LED + resistor pair should be connected to GND.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+use hal::gpio::DynPinId;
+use hal::gpio::FunctionSioOutput;
+use hal::gpio::Pin;
+use hal::gpio::PullDown;
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+// Alias for our HAL crate
+use rp2040_hal as hal;
+
+// A shorter alias for the Peripheral Access Crate, which provides low-level
+// register access
+use hal::pac;
+
+// Some traits we need
+use embedded_hal::blocking::delay::DelayMs;
+use embedded_hal::digital::v2::OutputPin;
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+/// Note: This boot block is not necessary when using a rp-hal based BSP
+/// as the BSPs already perform this step.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+/// Entry point to our bare-metal application.
+///
+/// The `#[rp2040_hal::entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables and the spinlock are initialised.
+///
+/// The function configures the RP2040 peripherals, then toggles a GPIO pin in
+/// an infinite loop. If there is an LED connected to that pin, it will blink.
+#[rp2040_hal::entry]
+fn main() -> ! {
+    // Grab our singleton objects
+    let mut pac = pac::Peripherals::take().unwrap();
+
+    // Set up the watchdog driver - needed by the clock setup code
+    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
+
+    // Configure the clocks
+    let clocks = hal::clocks::init_clocks_and_plls(
+        XTAL_FREQ_HZ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    // We will use the RP2040 timer peripheral as our delay source
+    let mut timer = rp2040_hal::Timer::new(pac.TIMER, &mut pac.RESETS, &clocks);
+
+    // The single-cycle I/O block controls our GPIO pins
+    let sio = hal::Sio::new(pac.SIO);
+
+    // Set the pins to their default state
+    let pins = hal::gpio::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // To put pins into an array we have to convert them to Dynamically Typed pins.
+    // This means they'll carry their pin and bank numbers around with them at run time,
+    // rather than relying on the Type of the pin to track that.
+    let mut pinarray: [Pin<DynPinId, FunctionSioOutput, PullDown>; 4] = [
```

Comment:
`PullNone` would be better for push-pull outputs.

Reply:
Agreed. But that's actually the value of what it's turned into by calling into_dyn_pin - not sure if that is unintentional, but it sure feels weird to have to change the pulls again after converting it.

Reply:
`PullDown` is the default pull mode, that what `pins.gpio2` starts with. Neither `into_push_pull_output()` nor `into_dyn_pin()` changes that.

Possible alternatives are:
- `pins.gpio2.into_push_pull_output().into_pull_type().into_dyn_pin()`
- `pins.gpio2.reconfigure().into_dyn_pin()`

Reply:
`.into_pull_type()` seems to have done the trick, but, once again, feels weird to have to say it .

Reply:
Why does it feel weird? Maybe unnecessarily verbose. But it is a nicely orthogonal API, where each of the `_into*` functions changes one of the generic parameters. If you want to change multiple of them at once, there's `reconfigure()`.

Reply:
> PullNone would be better for push-pull outputs.
> That's usually the best choice

This, basically, is what makes me feel a bit strange about it.
As you say, it's not like the Dyn behaviour is different. It's just that the base behaviour was unexpected.

> there's reconfigure().

True. I guess I could write it like
```rust
let mut pinarray: [Pin<DynPinId, FunctionSioOutput, PullNone>; 4] = [
        pins.gpio2.reconfigure().into_dyn_pin(),
        pins.gpio3.reconfigure().into_dyn_pin(),
        pins.gpio4.reconfigure().into_dyn_pin(),
        pins.gpio5.reconfigure().into_dyn_pin(),
    ];
```
or
```rust
let mut pinarray: [Pin<_, FunctionSioOutput, PullNone>; 4] = [
        pins.gpio2.reconfigure().into_dyn_pin(),
        pins.gpio3.reconfigure().into_dyn_pin(),
        pins.gpio4.reconfigure().into_dyn_pin(),
        pins.gpio5.reconfigure().into_dyn_pin(),
    ];
```

but while that makes it neater, it's now more magic (at least for someone new to the chip, the hal, and rust).

Reply:
I added another line showing use of reconfigure. I'm happy enough with this now.